### PR TITLE
Updated EXPM1() and Tests to New Engine

### PR DESCRIPTION
**Pull request by matthewryanwells**
Signed-off-by: Matthew Wells <matthew.wells@improving.com>

### Description
Updated EXPM1 function to run on the new engine, as well as added integration tests and unit tests
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 1]
```diff
@@ -197,6 +198,18 @@ private static DefaultFunctionResolver exp() {
...
+   * Definition of expm1(x) function. Calculate exponent function e to the x, minus 1
+   * The supported signature of exp function is INTEGER/LONG/FLOAT/DOUBLE -> DOUBLE
+   */
+  private static DefaultFunctionResolver expm1() {
+    return FunctionDSL.define(BuiltinFunctionName.EXPM1.getName(),
+        ExprCoreType.numberTypes().stream()
+            .map(type -> FunctionDSL.impl(FunctionDSL.nullMissingHandling(
+                v -> new ExprDoubleValue(Math.expm1(v.doubleValue()))),
+                type, DOUBLE)).collect(Collectors.toList()));
+  }
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:211]: Is it possible to extract common code piece to another function? It could be re-used across all other Math functions.

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java [hunk 2]
```diff
@@ -565,6 +565,77 @@ public void exp_missing_value() {
...
+            expm1.valueOf(valueEnv()),
+            allOf(hasType(DOUBLE), hasValue(Math.expm1(value))));
+    assertEquals(String.format("expm1(%s)", value.toString()), expm1.toString());
+  }
+
+  /**
+   * Test expm1 with null value.
+   */
+  @Test
+  public void expm1_null_value() {
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:624]: You can omit null/missing tests, because it is covered by `FunctionDSLnullMissingHandlingTest`
You don't actually test your function there, you test `nullMissingHandling`

**Review by Yury-Fridlyand — COMMENTED** [review]
What about PPL

**Comment by GumpacG** [comment]
> What about PPL?

https://github.com/opensearch-project/sql/pull/1097 for reference

**Comment by matthewryanwells** [comment]
> Actions are failing due to checkstyle.

the checkstyle was failing because of the lack of a newline which I have now fixed

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java (hunk 2 above)
Comment (matthewryanwells) [reply core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:624]: Would this mean that every function in `MathematicalFunctionTest.java` doesn't need a null/missing test? Because currently all math functions have tests for null/missing in the exact same format

**Comment by matthewryanwells** [comment]
> > What about PPL?
> 
> [opensearch-project#1097](https://github.com/opensearch-project/sql/pull/1097) for reference

I have added in PPL, including testing and documentation

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java (hunk 2 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:624]: Yes I believe we do not need it, even if it exists for other functions. I believe I had similar comments on a previous PR which included tests in `DateTimeFunctionTests.java` (IIRC the behaviour for `nullMissingHandling` is the same for all tests, so it doesn't need to be tested multiple times by each function).

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 1 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:211]: +1

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java (hunk 2 above)
Comment (matthewryanwells) [reply core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java]: I have removed the unnecessary tests for EXPM1

**Review by Yury-Fridlyand — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 3]
```diff
@@ -85,6 +86,14 @@ public static void register(BuiltinFunctionRepository repository) {
     repository.register(tan());
   }
 
+  private static DefaultFunctionResolver doubleMathFunctionBase(FunctionName functionName, SerializableFunction<Double, Double> formula) {
```
Comment (GumpacG) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:89]: I think it would be nice to have a Javadoc comment for this function

**Review by Yury-Fridlyand — APPROVED** [review]
You can rebase on `upstream/main` to make CI pass

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 3 above)
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java]: I have added a Javadoc comment

**Comment by matthewryanwells** [comment]
> You can rebase on `upstream/main` to make CI pass

I have now rebased the changes

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java [hunk 4]
```diff
@@ -565,6 +565,58 @@ public void exp_missing_value() {
     assertTrue(exp.valueOf(valueEnv()).isMissing());
   }
 
+  /**
+   * Test expm1 with integer value.
+   */
+  @ParameterizedTest(name = "expm1({0})")
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:571]: we should add tests for _at least_ each of the ExprCoreTypes:
```
          .add(BYTE)
          .add(SHORT)
          .add(INTEGER)
          .add(LONG)
          .add(FLOAT)
          .add(DOUBLE)
```

Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:571]: We may want to add a test in the file that ensures that  ExprCoreType.numberTypes() doesn't change (or if it does, someone needs to take a look at this test and updates the file with new test cases)

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java [hunk 5]
```diff
@@ -565,6 +565,58 @@ public void exp_missing_value() {
...
+            expm1.valueOf(valueEnv()),
+            allOf(hasType(DOUBLE), hasValue(Math.expm1(value))));
+    assertEquals(String.format("expm1(%s)", value.toString()), expm1.toString());
+  }
+
+  /**
+   * Test expm1 with float value.
+   */
+  @ParameterizedTest(name = "expm1({0})")
+  @ValueSource(floats = {-2F, 2F})
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java:598]: better to test using values that 'push the limits' of the data type
e.g. longs should be larger than the size of an INT, doubles should be larger than the siZe of floats, etc.

**Review by GabeFernandez310 — APPROVED** [review]

File: docs/user/dql/functions.rst [hunk 6]
```diff
@@ -443,10 +443,21 @@ EXPM1
...
 
-1. EXPM1(NUMBER T) -> T
+Return type: DOUBLE
 
+Example::
+
+    os> SELECT EXPM1(-1), EXPM1(0), EXPM1(1);
+    fetched rows / total rows = 1/1
+    +---------------------+------------+-------------------+
+    | EXPM1(-1)           | EXPM1(0)   | EXPM1(1)          |
```
Comment (acarbonetto) [inline docs/user/dql/functions.rst:457]: use a double/float as an example

File: docs/user/ppl/functions/math.rst [hunk 7]
```diff
@@ -321,6 +321,27 @@ Example::
...
+
+Usage: expm1(x) return e raised to the power of x, minus 1.
+
+Argument type: INTEGER/LONG/FLOAT/DOUBLE
+
+Return type: DOUBLE
+
+Example::
+
+    os> source=people | eval `EXPM1(2)` = EXPM1(2) | fields `EXPM1(2)`
```
Comment (acarbonetto) [inline docs/user/ppl/functions/math.rst:338]: use a double/float as an example

**Review by acarbonetto — CHANGES_REQUESTED** [review]

**Comment by GumpacG** [comment]
I would suggest to have the math function base extraction be a part of a different task so that more tests can be included for each of the types that the functions supports. This would prevent the task and PR from becoming large.

**Review by forestmvey — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 8]
```diff
@@ -85,6 +86,22 @@ public static void register(BuiltinFunctionRepository repository) {
...
+   * @param formula lambda function of math formula.
+   * @return DefaultFunctionResolver for math functions.
+   */
+  private static DefaultFunctionResolver doubleMathFunctionBase(
+          FunctionName functionName, SerializableFunction<Double, Double> formula) {
+    return FunctionDSL.define(functionName,
+        ExprCoreType.numberTypes().stream()
+            .map(type -> FunctionDSL.impl(FunctionDSL.nullMissingHandling(
+                    v -> new ExprDoubleValue(formula.apply(v.doubleValue()))),
+                DOUBLE, type)).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java]: I changed `DOUBLE` to being a passed in parameter

File: core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java (hunk 5 above)
Comment (matthewryanwells) [reply core/src/test/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunctionTest.java]: I have improved the quality of the tests to include edge cases

File: docs/user/dql/functions.rst (hunk 6 above)
Comment (matthewryanwells) [reply docs/user/dql/functions.rst]: I included `EXPM1(1.5)` as a test case

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 9]
```diff
@@ -85,6 +86,23 @@ public static void register(BuiltinFunctionRepository repository) {
     repository.register(tan());
   }
 
+  /**
+   * Base function for math functions with similar formats that return DOUBLE.
+   *
+   * @param functionName BuiltinFunctionName of math function.
+   * @param formula lambda function of math formula.
+   * @return DefaultFunctionResolver for math functions.
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:94]: Reflect 3rd arg in javadoc.

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 10]
```diff
@@ -85,6 +86,23 @@ public static void register(BuiltinFunctionRepository repository) {
...
+   */
+  private static DefaultFunctionResolver doubleMathFunctionBase(
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:96]: 1. You can rename this function now
2. You can use it for `sign`, `sqrt` and  `log(x)` functions too

File: docs/user/ppl/functions/math.rst [hunk 11]
```diff
@@ -321,7 +321,6 @@ Example::
     | 7.38905609893065 |
     +------------------+
 
-
```
Comment (GumpacG) [inline docs/user/ppl/functions/math.rst:324]: Accidental deletion?

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 12]
```diff
@@ -158,7 +177,7 @@ private static DefaultFunctionResolver conv() {
 
   /**
    * Definition of crc32(x) function.
-   * Calculate a cyclic redundancy check value and returns a 32-bit unsigned value
+   * Calculate a cyclic redundancy check value and returns a 32-bi t unsigned value
```
Comment (GumpacG) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:180]: Accident?

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 13]
```diff
@@ -181,20 +200,26 @@ private static DefaultFunctionResolver crc32() {
    */
   private static DefaultFunctionResolver euler() {
     return FunctionDSL.define(BuiltinFunctionName.E.getName(),
-        FunctionDSL.impl(() -> new ExprDoubleValue(Math.E), DOUBLE)
+            FunctionDSL.impl(() -> new ExprDoubleValue(Math.E), DOUBLE)
```
Comment (GumpacG) [inline core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:203]: Why did this function get reverted? Shouldn't it work with the base math function?

Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:203]: The function was reverted because the format is different than that of other math functions, I believe it is because of the fact that euler doesn't take in any parameters so it is missing a section of code that all of the other functions have

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 12 above)
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:180]: yes, I will fix it now

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 13 above)
Comment (GumpacG) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java:203]: If that's the case, shouldn't the base function be able to handle it? Maybe have an optional parameter or a null can be passed it.

**Review by Yury-Fridlyand — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 13 above)
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java]: The functions are so fundamentally different that a different base function would be required to be written for them, and as the only functions that would be able to use that new base function would be E and Pi I feel that it is currently unnecessary to do currently

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java [hunk 14]
```diff
@@ -85,6 +87,23 @@ public static void register(BuiltinFunctionRepository repository) {
     repository.register(tan());
   }
 
+  /**
+   * Base function for math functions with similar formats that return DOUBLE.
+   *
+   * @param functionName BuiltinFunctionName of math function.
+   * @param formula lambda function of math formula.
+   * @param dataType data type input for the function
```
[the remark this answers is not in the document]
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java]: I've updated the function and javadoc to mention that it is the returnType and not input

File: core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java (hunk 10 above)
Comment (matthewryanwells) [reply core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/MathematicalFunction.java]: `sign` and `sqrt` now use the base function, `log(x)` format is too different to be included
